BUG: Some sas7bdat files with many columns are not parseable by read_sas

troels opened the pull request:
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

The reason is that column definitions may be split up into different pages.
Allow column information to be parsed from different pages
and add a test for it.

troels committed:
BUG: Some sas7bdat files with many columns are not parseable by read_sas

The reason is that column definitions may be split up into different pages.
Allow column information to be parsed from different pages
and add a test for it.

pep8speaks commented:
Hello @troels! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/io/sas/sas7bdat.py`](https://github.com/troels/pandas/blob/f36889598b2d847ce3f83f5464f22bdb1fbef01e/pandas/io/sas/sas7bdat.py) !

 - There are no PEP8 issues in the file [`pandas/tests/io/sas/test_sas7bdat.py`](https://github.com/troels/pandas/blob/f36889598b2d847ce3f83f5464f22bdb1fbef01e/pandas/tests/io/sas/test_sas7bdat.py) !

gfyoung commented:
@troels : Thanks for the report!  Do you mind sharing in this PR what output you get without your patch for reference?  A small, reproducible example would be great.

gfyoung commented on pandas/tests/io/sas/test_sas7bdat.py:
```diff
@@ -183,6 +183,22 @@ def test_date_time(datapath):
+                          'ecgrtxt': np.dtype('S'),
+                          'ecgrtime': np.dtype('d'),
+                          'nvitl1': np.dtype('d')
+                      })
```
Generally, we would prefer to construct the expected DataFrame from scratch.  This is one of the motivations for having a small, reproducible example.

troels replied:
Hi @gfyoung, I'd generally agree, but in this case, I don't have a good way of constructing another test example. I could perhaps use a more type-safe format than csv, as the reference file, do you have any good ideas?

gfyoung replied:
Gotcha.  Let's see what you have here: one header and three rows IIUC.  Do you need that many columns to construct a breaking example?

troels replied:
Yeah. The problem occurs as a consequence of the number of columns. Otherwise SAS will save all column info on one page, and the problem will not be replicated.

gfyoung replied:
Hmm...I see.  Alright, let's leave this as is for now then. 👍

gfyoung commented on doc/source/whatsnew/v0.23.5.txt:
```diff
@@ -49,3 +49,5 @@ Bug Fixes
 
 - Bug in :func:`read_csv` that caused it to raise ``OverflowError`` when trying to use 'inf' as ``na_value`` with integer index column (:issue:`17128`)
+
+- Bug in :func:`read_sas` that caused it to fail parsing some sas7bdat files with many columns
```
* This should go in `0.24.0`
* Reference your PR as the issue number

gfyoung commented on pandas/io/sas/sas7bdat.py:
```diff
@@ -102,6 +103,15 @@ def __init__(self, path_or_buf, index=None, convert_dates=True,
+        return np.asarray(self._column_data_offsets, dtype=np.int64)
+
+    def column_types(self):
+        return np.asarray(self._column_types, dtype=np.dtype('S1'))
```
Can we have docstrings for these methods?

troels commented:
Hi @gfyoung 

So the test case I added will fail in the current version of pandas, with the following error (these files can't be parsed):

pandas/tests/io/sas/test_sas7bdat.py ............F.                                              [100%]

=============================================== FAILURES ===============================================
__________________________________________ test_many_columns ___________________________________________

datapath = <function datapath.<locals>.deco at 0x7f53f5a47f28>

    def test_many_columns(datapath):
        fname = datapath("io", "sas", "data", "many_columns.sas7bdat")
>       df = pd.read_sas(fname, encoding='latin-1')

pandas/tests/io/sas/test_sas7bdat.py:188: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas/io/sas/sasreader.py:68: in read_sas
    data = reader.read()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pandas.io.sas.sas7bdat.SAS7BDATReader object at 0x7f53f5899860>, nrows = 3

    def read(self, nrows=None):
    
        if (nrows is None) and (self.chunksize is not None):
            nrows = self.chunksize
        elif nrows is None:
            nrows = self.row_count
    
        if len(self.column_types) == 0:
            self.close()
>           raise EmptyDataError("No columns to parse from file")
E           pandas.errors.EmptyDataError: No columns to parse from file

pandas/io/sas/sas7bdat.py:604: EmptyDataError
================================= 1 failed, 13 passed in 18.88 seconds =================================

troels committed:
BUG: Fix parsing of sas7bdat files with odd data pages (#16615)

SAS can apparently generate data pages having bit 7 (128) set on
the page type.
It seems that the presence of bit 8 (256) determines whether it's
a data page or not. So treat page as a data page if bit 8 is set and
don't mind the lower bits.

troels commented:
Hi @gfyoung 

I added a commit fixing #16615 too. I hope it's ok to continue using the same PR, as the two pieces of code build upon each other.

gfyoung commented:
> I added a commit fixing #16615 too. I hope it's ok to continue using the same PR, as the two pieces of code build upon each other.

@troels : In this case, it should be fine.  Thanks for letting me know!

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -690,7 +690,8 @@ I/O
 - :func:`read_csv()` will correctly parse timezone-aware datetimes (:issue:`22256`)
--
+- :func:`read_sas` will correctly parse sas7bdat files with many columns (:issue:`22628`)
+- :func:`read_sas` will correctly parse sas7bdat files with odd data page types (:issue:`16615`)
```
can you expand on 'odd' here?

jreback commented on pandas/io/sas/sas.pyx:
```diff
@@ -437,7 +437,7 @@ cdef class Parser(object):
                 # string
                 string_chunk[js, current_row] = np.array(source[start:(
-                    start + lngt)]).tostring().rstrip()
+                    start + lngt)]).tostring().rstrip(b"\x00 ")
```
there is a space here as well?

troels replied:
Hi @jreback 

Yes, there is supposed to be a space, just like many other places in sas7bdat.py. 

Apparently SAS can both pad with zeros and spaces to the right. (Maybe depending on version or something else).

jreback commented on pandas/tests/io/sas/test_sas7bdat.py:
```diff
@@ -183,6 +183,28 @@ def test_date_time(datapath):
+    tm.assert_frame_equal(df, df0)
+
+
+def test_inconsistent_number_of_rows(datapath):
```
can you add the issue number as a comment

jreback commented on pandas/tests/io/sas/test_sas7bdat.py:
```diff
@@ -183,6 +183,28 @@ def test_date_time(datapath):
 
 
+def test_many_columns(datapath):
+    fname = datapath("io", "sas", "data", "many_columns.sas7bdat")
```
same

troels commented:
The test failure here is quite unrelated to the PR.

troels commented:
Hi @jreback 

Anything else preventing this from being merged?

gfyoung reviewed (approved):
I'm not @jreback , but I don't think there are any barriers to merging this.  LGTM!

jreback commented:
thanks!

troels commented:
Thanks both of you :)
